jsonstar/JsonStar.Schema.Ast.Types.fst

```diff
@@ -0,0 +1,128 @@
+/// This module provides a simplified view of F* terms we want to translate into json-schema
+module JsonStar.Schema.Ast.Types
+
+// ..\packages\FStar.Windows.Ocaml.Unofficial.0.4.0\tools\bin\fstar.exe  --error_contexts true --odir obj\Debug\extracted --codegen FSharp --use_hints --record_hints JsonStar.Schema.Ast.Types.fst --extract "+JsonStar.Schema.Ast.Types"
+
+let description (desc : string) : Tot unit = ()
+let defaultValue (v : string) : Tot unit = ()
+let readOnly : unit = ()
+/// Provided value has to be known at compile-time
+let readOnlyIf (v : bool) : Tot unit = ()
+
+type format_type = 
+    | Percentage
+    | Date
+
+let format_type_to_string (x : format_type) = 
+    match x with
+    | Percentage -> "percentage"
+    | Date       -> "date"
+
+let format (f : format_type) : Tot unit = ()
+
+/// Attributes that can be associated with anything that represents a field in json-schema
+type value_attributes = 
+    {
+        _description  : option string;
+        _defaultValue : option string;
+        _readOnly     : option bool;
+        _format       : option format_type;
+    }
+
+/// Attributes that can be associated with anything that represents an object in json-schema
+type object_attributes =
+    {
+        _description : option string;
+        // Id, schemaVersion, title, AllowAdditionalProperties ... (?)
+    }
+
+type number_refinement = 
+    | Minimum : v:int -> number_refinement
+    | Maximum : v:int -> number_refinement
+
+type string_refinement = 
+    | MinLength : v:nat -> string_refinement
+    | MaxLength : v:nat -> string_refinement
+    | Pattern : v:string -> string_refinement
+
+type enum_refinement = 
+    | Allow : v:list string -> enum_refinement
+    | Disallow : v:list string -> enum_refinement
```

Comment:
It's true. On the other hand the code here is not meant to be used directly by the user (this would allow cheating), but rather it's meant as a wrapper over F* terms, and the representation is generated from real F* types. Still you're right that a better representation could prevent some bugs like e.g. accidentally merging refinements for two different enums.